### PR TITLE
Build optimizations, dev deploy updates

### LINKED ISSUE #102: dev-deployment dead
as of #84, builds began failing because out of memory errors

```
❯ inertia nwdev logs nwhacks2019
/2018-09-07T22:11:40.671513464Z yarn run v1.6.0
y2018-09-07T22:11:40.794896493Z $ cross-env NODE_ENV=development webpack-dev-server --hot --inline --watch --host 0.0.0.0
l2018-09-07T22:11:50.023322750Z ℹ ｢wds｣: Project is running at http://0.0.0.0:8080/
a2018-09-07T22:11:50.026516499Z ℹ ｢wds｣: webpack output is served from /
c2018-09-07T22:11:50.029396928Z ℹ ｢wds｣: 404s will fallback to /index.html
 2018-09-07T22:12:10.340331753Z
 2018-09-07T22:12:10.523053720Z
!2018-09-07T22:12:10.523064655Z #
92018-09-07T22:12:10.523068650Z # Fatal error in , line 0
^2018-09-07T22:12:10.523072320Z # API fatal error handler returned after process out of memory
!2018-09-07T22:12:10.523076022Z #
t2018-09-07T22:12:10.762835513Z info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
F2018-09-07T22:12:10.793833944Z error Command failed with exit code 1.
FATA[0028] websocket: close 1006 (abnormal closure): unexpected EOF
```

### LINKED ISSUE #103: for photos, use `png` and compress them instead of `svg`
`olivia.svg` for e.g. is > 11MB (yikes) so use `png` and compress

promising suspect for #102

### PR DESCRIPTION
**Pull request by bobheadxi**
:tickets: **Ticket(s)**: Closes #102, closes #103 

---

### Optimizations
- photo `svg` assets have been converted to `png` for a pretty big size reduction (10+mb to a few hundred kb)
- some bundle splitting implemented - our main bundle is still kinda large, but they're better spread out now

### Dev Tooling
- new `FIREBASE` env variable to toggle prod and dev Firebase credentials
- some dependency recategorizing
- `.env` is loaded in build only for dev-server
- bundleanalyzer for bundle size analysis with `yarn start` 
image

## :thought_balloon: Notes

My slack message:

```
so I can't seem to run the builds on the free ec2 instance's 1gb of ram. i've made some changes and will put up a PR tonight based on this plan instead:
- we set up 2 branches, `master` and `dev`. `dev` will be our main branch, all changes except hotfixes will go here first
- the ec2 instance will use whatever is in `/docs` on `dev` and deploy it using dev credentials
- gh-pages will use whatever is in `/docs`  on `master` and deploy it using prod credentials
- builds need to be committed to have changes actually appear in either `dev` or `master`
downsides:
- no more automated build :disappointed: we could set up travis to automatically build and commit stuff but.... idk about that
upsides:
- we were previously leaking credentials in `bundle.js` because of the use of `.env` loader inside the actual codebase, this has been fixed and hopefully no one will notice
- with the above change, the static web app is actually configurable from env (not that it matters for gh-pages)
```

## :flashlight: Testing Instructions

```
make web
```

### CODE REVIEW

File: web/components/app/App.js [hunk 1]
```diff
@@ -3,20 +3,31 @@ import PropTypes from 'prop-types';
...
+});
+const Logout = asyncComponent(async () => {
+  const auth = await import('../auth');
+  return auth.Logout;
+});
+const HackerApplication = asyncComponent(async () => {
+  const application = await import('../../containers/application');
+  return application.HackerApplication;
+});
+const AdminPanel = asyncComponent(() => import('../admin'));
```
Comment (mingyokim) [inline web/components/app/App.js]: why is this needed?

**Comment by mingyokim** [comment]
also `/application/hacker` and `/login` pages are blank?

File: web/components/app/App.js (hunk 1 above)
Comment (bobheadxi) [reply web/components/app/App.js]: Asynchronous imports allow you to use multiple bundles, users don’t need the JavaScript downloaded for admin panel for example until they visit it, hence why the docs folder now has a bunch of things

File: web.Dockerfile [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
-COPY web/package.json web/package.json
-COPY web/yarn.lock web/yarn.lock
+COPY docs docs
 
-WORKDIR web
-RUN yarn install --production --silent
-COPY web .
+RUN yarn global add http-server
+
+RUN echo "source ./.env ; export NODE_ENV=development ; http-server ./docs -a 0.0.0.0 -p 8080 -c-1" > entry.sh
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply web.Dockerfile]: Was trying to reduce memory usage - this is part of the change to serve the static site instead of building on ec2

[the remark this answers is not in the document]
Comment (bobheadxi) [reply web.Dockerfile]: The dockerfile is used for ec2

**Comment by bobheadxi** [comment]
I’ll investigate the issues hm

File: web/components/app/App.js (hunk 1 above)
Comment (bobheadxi) [reply web/components/app/App.js]: Also, see the diagram in the pr description for the breakdown

**Comment by bobheadxi** [comment]
@mingyokim i believe ive fixed the blank pages in ffe3a86

File: web.Dockerfile (hunk 2 above)
Comment (chamkank) [inline web.Dockerfile]: I like how you piped this, saves us from having another file in the main directory 💯

File: web/webpack.plugins.js [hunk 3]
```diff
@@ -1,7 +1,6 @@
 const webpack = require('webpack');
 const MinifyPlugin = require('babel-minify-webpack-plugin');
 const HtmlWebpackPlugin = require('html-webpack-plugin');
-const Dotenv = require('dotenv-webpack');
 
 // Define webpack plugins used in dev and prod configurations here
```
Comment (chamkank) [inline web/webpack.plugins.js]: Can you explain why Dotenv was moved from `webpack.plugins.js` to `webpack.config.js`?

Comment (bobheadxi) [reply web/webpack.plugins.js]: To keep dev config hidden - builds use `webpack.plugins.js` and `webpack.prod.js`, the dev-server uses `webpack.config.js`. This prevents us from leaking credentials in .env into build bundbles

Comment (chamkank) [reply web/webpack.plugins.js]: Got it :ok_hand: In that case we should probably update the comment on line 5 of `webpack.plugins.js`. Could also make it extra clear by renaming `webpack.config.js` to `webpack.dev.js`.

Comment (bobheadxi) [reply web/webpack.plugins.js]: doesn't line 5 already say something along those lines?

```
// Define webpack plugins used in dev and prod configurations here
```

Comment (chamkank) [reply web/webpack.plugins.js]: Oh my bad, is that file used for both the prod build and the dev server? If so, no changes needed
